### ISSUE TITLE
[Bug]: Unexpected build does not build docker images

### ISSUE THREAD

**Issue by ghostsquad** [goreleaser/goreleaser#2710]
### What happened?

```
goreleaser build --snapshot --rm-dist
```

builds my go binaries, but not my docker image

and

```
goreleaser release --snapshot --rm-dist
```

builds my docker images

---

from a developer experience standpoint, I found it unexpected that `goreleaser` did not build the docker images as part of the `build` phase. As I'm attempting to simplify my processes for building the binary and docker image, but also maintain unification between local development and my automated CI/CD flow. I'd like to build _all_ relevant artifacts, review/test, etc before releasing. As it is right now, I cannot build the docker image(s) without running `release`

### How can we reproduce this?

```
goreleaser build --snapshot --rm-dist
```

### goreleaser version

```bash
goreleaser version 1.1.0
commit: af4a8642bddb171914ee932bfb5768f102c3516f
built at: 2021-11-28T23:41:04Z
built by: goreleaser
module version: v1.1.0, checksum: h1:YySqqYTX9kxRU0e/fGQrhivXk8/zD4iUOlL7h0I2gBk=

https://goreleaser.com
```

### GoReleaser Check

- [X] goreleaser check shows no errors

### Search

- [X] I did search for other open and closed issues before opening this.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by caarlos0** [comment]
this is by design, not sure if we're going to allow to build the docker images as well...

if that's what you need, probably having a second goreleaser config just for that is better...

**Comment by ghostsquad** [comment]
ya, I spent the afternoon learning about goreleaser for the first time. It's been on my radar for quite a while, and with 1.0.0 released just 2 weeks ago, I thought I would look at this again, and introduce it to my company.

As I worked through the documentation, I realized it took me awhile to develop a mental model of the steps the goreleaser was going to do, and specifically maybe the terms "build", "release" and "publish". With "release" and "publish" almost being synonyms, I struggled to understand what was actually different, and how I would create local development workflows, and how I would create my CI pipelines.

I was thinking maybe this is just a failing of the documentation online to help me develop the mental model of the tool.

Here's where it kind of began:

https://goreleaser.com/how-it-works/

```
## What happens when you release

A GoReleaser run is split into 4 major steps:

- defaulting: configures sensible defaults for each step
- building: builds the binaries, archives, packages, Docker images, etc
- publishing: publishes the release to the configured SCM, Docker registries, blob storages...
- announcing: announces your release to the configured channels
```

so, I read step 2, and then saw `goreleaser build` subcommand, and thought those were matched. This is not the case.

Then there's `goreleaser release` but no `goreleaser publish`, instead `goreleaser --release` has a `--skip-publish` flag.

From a fresh eyes perspective, I found it a bit confusing, and after a couple hours of playing around, I felt it relevant to at least create this issue and start a conversation about it.

**Comment by ghostsquad** [comment]
There's no `publish` subsection in the config, instead it's just a series of individual keys based on the publishing _method_, such as a "github `release`"

`announce` though is itself a section, with subkeys for the different ways to produce the announcement. That makes sense.

**Comment by caarlos0** [comment]
you're right, yeah, docs on that manner should be better... the idea behind `goreleaser build` was to replace the `go build` stuff in most makefiles...

Announce has its own section because it makes sense for it, but some of the others mix a bit, e.g. docker section setups both docker image builds and publishing, and yeah, 100% it can be confusing.

About the specific case of `goreleaser build`, will improve its docs and `--help` so it hopefully makes a bit more sense.

**Comment by ghostsquad** [comment]
@caarlos0 thank you for helping me out. I think it helped that I just wrote this down in an issue. Creating an accurate mental model of something sometimes takes several steps:

1. read the docs
2. try it yourself
3. try to explain it to someone else (in this case, in the form of a github issue)

so ya... it took me all the way to step 3 to really nail this down!

Is there a way I can contribute to docs? maybe I could write my own, and show you somehow?

**Comment by caarlos0** [comment]
> Is there a way I can contribute to docs? maybe I could write my own, and show you somehow?

sure thing! For the `--help` itself, you can edit [here](https://github.com/goreleaser/goreleaser/blob/main/cmd/build.go#L46) (the markdown files are auto-generated from it, so you don't need to worry about it).

Other pages you might want to edit are within the `www` folder: https://github.com/goreleaser/goreleaser/tree/main/www/docs

You can use `task docs:serve` to preview it locally (needs docker or podman) as well.

**Comment by caarlos0** [comment]
fwiw, if you need more guidance and rather chat about it, we have a channel on gophers slack and also a discord channel (discord is kind of preferred as we have more control there): https://goreleaser.com/links/

**Comment by ghostsquad** [comment]
Within this vein, I imagine that `changelog` is kind of like an artifact that is "published" as part of a "release".

I have the following in my config:

```
changelog:
  sort: asc
  filters:
    exclude:
      - '^docs:'
      - '^test:'
```

I'd like to see what the Changelog would look like. I assumed I'd be able to do that (with the new mental model of using `goreleaser release --rm-dist --snapshot --skip-publish`) but once again, I'm not sure if a changelog is actually being generated or where I can find it to review.

The documentation also doesn't explain what is being compared when it comes to generating the changelog, is it based on git tags? What versions am I comparing for the changelog generation?

**Comment by caarlos0** [comment]
Almost! 

`--snapshot` skips the changelog generation, and `--skip-publish` is implied by `--snapshot`.

But you can run `release --rm-dist --skip-validate --skip-publish`, and that should result in you having a `dist/CHANGELOG.md`. You can see its generated early in the process, so you can cancel the workflow once you see the logs if you don't want to wait for the whole thing...

**Comment by caarlos0** [comment]
will close as docs are improved a bit, @ghostsquad feel free to open PRs with things you find could be better, if you still have any 🤘

Thanks!
